quine-cypher/src/main/scala/com/thatdot/quine/compiler/cypher/QueryPart.scala

```diff
@@ -12,29 +12,54 @@ object QueryPart {
     *
     * @param queryPart query to compiler
     * @param isEntireQuery this query part is the whole query
+    * @param isSubQuery is this inside a `CALL { .. }`?
     * @return execution instructions for Quine
     */
   def compile(
     queryPart: ast.QueryPart,
-    isEntireQuery: Boolean = true
+    isEntireQuery: Boolean = true,
+    isSubQuery: Boolean = false
   ): CompM[cypher.Query[cypher.Location.Anywhere]] =
     queryPart match {
-      case ast.SingleQuery(clauses) =>
-        compileClauses(clauses, isEntireQuery)
+      case sq: ast.SingleQuery =>
+        if (!isSubQuery) {
+          compileClauses(sq.clauses, isEntireQuery)
+        } else {
+          for {
+            // Prepare for the subquery to run by setting the imported columns
+            initialColumns: Vector[Symbol] <- CompM.getColumns
+            importedVariables =
+              if (sq.isCorrelated) {
+                sq.importColumns.view.map(Symbol.apply).toVector
+              } else {
+                initialColumns
+              }
+            () <- CompM.clearColumns
+            () <- importedVariables.traverse_(CompM.addColumn)
+
+            // Compile the subquery
+            subQuery <- compileClauses(sq.clausesExceptImportWith, isEntireQuery)
+
+            // Update the columns by appending back all of the initial columns
+            () <- initialColumns.traverse_(CompM.addColumn)
```

Comment:
We know that the `initialColumns` will be distinct from `importedVariables`? At least in the case of `isCorrelated == false`, those will be the same thing - meaning that same set will get added twice - and I'm used to `addColumn` throwing an exception from its `require` statement if you try to add duplicates. Or does `compileClauses` clear that out in between or something?

Reply:
> We know that the `initialColumns` will be distinct from `importedVariables`

No. on the contrary, they can be the same (I think that's what happens if you don't explicitly use `WITH`, and I believe that is the `isCorrelated = false` case)

> At least in the case of `isCorrelated == false`, those will be the same thing - meaning that same set will get added twice - and I'm used to `addColumn` throwing an exception from its `require` statement if you try to add duplicates.

Cypher doesn't let you return from a subquery any columns which would collide with those from the outer query. That means that we know for sure that after having compiled `subQuery`, none of the `intialColumns` will be in the current set of columns anymore.